katello/hooks/pre_commit/13-hiera.rb

```diff
@@ -1,17 +1,44 @@
-TUNING_SIZES = ['default'] + ['medium', 'large', 'extra-large', 'extra-extra-large']
+# A mapping of tuning profile to the required CPU cores and memory in GB
+# TODO: check actual required values
+TUNING_SIZES = {
+  'default' => [1, 8],
+  'medium' => [1, 10],
+  'large' => [1, 20],
+  'extra-large' => [1, 128],
+  'extra-extra-large' => [1, 256],
+}
 TUNING_FACT = 'tuning'
 
-current = get_custom_fact(TUNING_FACT)
-new = app_value(:tuning)
+current_tuning = get_custom_fact(TUNING_FACT)
+new_tuning = app_value(:tuning)
 
-if current != new
-  unless TUNING_SIZES.include?(new)
-    say "<%= color('Invalid tuning profile', :bad) %>"
-    say "'#{new}' is not one of #{TUNING_SIZES.join(', ')}"
-    exit 101
+required = TUNING_SIZES[new_tuning]
+if required.nil?
+  say "<%= color('Invalid tuning profile', :bad) %>"
+  say "'#{new_tuning}' is not one of #{TUNING_SIZES.keys.join(', ')}"
+  exit 101
+end
+
+unless app_value(:disable_system_checks)
+  required_cores, required_memory = required
+
+  # Check if it's actually 90% of the required. If a crash kernel is enabled
+  # then the reported total memory is lower than in reality.
+  if facts[:memory][:system][:total_bytes] < (required_memory * 1024 * 1024 * 1024 * 0.9)
+    say "<%= color('Insufficient memory for tuning size', :bad) %>"
+    say "Tuning profile '#{new_tuning}' requires at least #{required_memory} GB of memory"
+    exit 102
   end
 
-  store_custom_fact(TUNING_FACT, new)
+  if facts[:processors][:count] < required_cores
+    say "<%= color('Insufficient CPU cores for tuning size', :bad) %>"
+    say "Tuning profile '#{new_tuning}' requires at least #{required_cores} CPU cores"
+    exit 103
```

Comment:
What are the various exit numbers meanings?

Reply:
Well, 103 now means you have insufficient CPU cores :) Perhaps we should add a table in README similar to https://github.com/theforeman/kafo#exit-code

Reply:
That would be a good start. Mapping them to constants could also be a good tactic:

```
INSUFFICIENT_CPU_CORES = 103

exit INSUFFICIENT_CPU_CORES
```

Reply:
THoughts on this idea? Even if stored at the top of the file. I know that if I com eback to this file I will likely look at 101, 102, 103 and start googling Ruby exit codes and rabbit hole.

Reply:
I'll start with just defining them in the same file. We can do the general exit code cleanup. Then we should also look at the other hooks which probably override Kafo exit codes somewhere.

Reply:
Updated